File: app/src/main/java/com/kaciula/archiman/data/local/PrefsModule.java
package com.kaciula.archiman.data.local;

import android.content.Context;
import android.content.SharedPreferences;

import com.kaciula.archiman.data.local.prefs.BooleanPreference;
import com.kaciula.archiman.data.local.prefs.IntPreference;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class PrefsModule {

    @Provides
    @Singleton
    SharedPreferences provideSharedPreferences(Context ctx) {
        return ctx.getSharedPreferences("prefs_utils", Context.MODE_PRIVATE);
    }

    @Provides
    @Named("first_time")
    @Singleton
    BooleanPreference provideFirstTimePreference(SharedPreferences preferences) {
        return new BooleanPreference(preferences, "first_time", true);
    }

    @Provides
    @Named("version_code")
    @Singleton
    IntPreference provideVersionCodePreference(SharedPreferences preferences) {
        return new IntPreference(preferences, "version_code", 0);
    }
}

File: app/src/main/kotlin/com/kaciula/archiman/infrastructure/injection/PrefsModule.kt
package com.kaciula.archiman.infrastructure.injection

import android.content.Context
import android.content.SharedPreferences
import com.kaciula.archiman.infrastructure.data.local.prefs.util.BooleanPreference
import com.kaciula.archiman.infrastructure.data.local.prefs.util.IntPreference
import dagger.Module
import dagger.Provides
import javax.inject.Named
import javax.inject.Singleton

@Module
class PrefsModule {

    @Provides
    @Singleton
    fun provideSharedPreferences(ctx: Context): SharedPreferences {
        return ctx.getSharedPreferences("prefs_utils", Context.MODE_PRIVATE)
    }

    @Provides
    @Named("first_time")
    @Singleton
    fun provideFirstTimePreference(preferences: SharedPreferences): BooleanPreference {
        return BooleanPreference(preferences, "first_time", true)
    }

    @Provides
    @Named("version_code")
    @Singleton
    fun provideVersionCodePreference(preferences: SharedPreferences): IntPreference {
        return IntPreference(preferences, "version_code", 0)
    }
}
